src/libraries/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceController.cs

```diff
@@ -7,965 +7,1000 @@
-      IntPtr[] argPtrs = new IntPtr[args.Length];
-      int i = 0;
-      try
-      {
-        for (i = 0; i < args.Length; i++)
-        {
-          if (args[i] == null)
-            throw new ArgumentNullException($"{nameof(args)}[{i}]", SR.ArgsCantBeNull);
-
-          argPtrs[i] = Marshal.StringToHGlobalUni(args[i]);
-        }
-      }
-      catch
-      {
-        for (int j = 0; j < i; j++)
-          Marshal.FreeHGlobal(argPtrs[i]);
-        throw;
-      }
-
-      GCHandle argPtrsHandle = default;
-      try
-      {
-        argPtrsHandle = GCHandle.Alloc(argPtrs, GCHandleType.Pinned);
-        bool result = Interop.Advapi32.StartService(serviceHandle, args.Length, argPtrsHandle.AddrOfPinnedObject());
-        if (!result)
-        {
-          Exception inner = new Win32Exception(Marshal.GetLastWin32Error());
-          throw new InvalidOperationException(SR.Format(SR.CannotStart, ServiceName, _machineName), inner);
-        }
-      }
-      finally
-      {
-        for (i = 0; i < args.Length; i++)
-          Marshal.FreeHGlobal(argPtrs[i]);
-        if (argPtrsHandle.IsAllocated)
-          argPtrsHandle.Free();
-      }
-    }
+            int error = Interop.Advapi32.NotifyServiceStatusChange(serviceHandle, notifyMask, notify);
+            if (error != Interop.Errors.ERROR_SUCCESS)
+            {
+                throw new Win32Exception(error);
+            }
 
-    /// <summary>
-    /// Stops the service. If any other services depend on this one for operation,
-    /// they will be stopped first. The DependentServices property lists this set
-    /// of services.
-    /// </summary>
-    public unsafe void Stop()
-    {
-      using SafeServiceHandle serviceHandle = GetServiceHandle(Interop.Advapi32.ServiceOptions.SERVICE_STOP);
-      // Before stopping this service, stop all the dependent services that are running.
-      // (It's OK not to cache the result of getting the DependentServices property because it caches on its own.)
-      for (int i = 0; i < DependentServices.Length; i++)
-      {
-        ServiceController currentDependent = DependentServices[i];
-        currentDependent.Refresh();
-        if (currentDependent.Status != ServiceControllerStatus.Stopped)
-        {
-          currentDependent.Stop();
-          currentDependent.WaitForStatus(ServiceControllerStatus.Stopped, new TimeSpan(0, 0, 30));
-        }
-      }
-
-      Interop.Advapi32.SERVICE_STATUS status = default;
-      bool result = Interop.Advapi32.ControlService(serviceHandle, Interop.Advapi32.ControlOptions.CONTROL_STOP, &status);
-      if (!result)
-      {
-        Exception inner = new Win32Exception(Marshal.GetLastWin32Error());
-        throw new InvalidOperationException(SR.Format(SR.StopService, ServiceName, _machineName), inner);
-      }
-    }
+            uint status = Interop.Kernel32.SleepEx(millisecondsTimeout, true);
```

Comment:
If SleepEx returns WAIT_IO_COMPLETION, then that means the sleep was interrupted by an asynchronous procedure call, but I don't think you can assume that the APC was caused by NotifyServiceStatusChange. It might have been caused by some I/O that was started before WaitForStatus was called.

Reply:
That could then lead to stack corruption if WaitForStatus returns too early and frees the stack frame that contains the SERVICE_NOTIFY structure, and the actual status change occurs later and Windows updates fields of the freed structure.

Reply:
The stack corruption might usually be prevented because WaitForStatus disposes of serviceHandle before returning, and NotifyServiceStatusChange documentation says that closing the handle cancels the notification request. However, if the notification request queues an asynchronous procedure call immediately after SleepEx returns, I'm not sure whether CloseServiceHandle will cancel the APC as well. (There is QueueUserAPC  but AFAIK no public function for canceling an APC, so there might be no private function for that either.)

Reply:
> It might have been caused by some I/O that was started before WaitForStatus was called.

Any suggestion on this? I guess we should start a new native thread.

Reply:
For APCs caused by unrelated I/O, I guess an easy workaround would be to recheck the status of the service after the sleep. Alternatively, embed the notification structure in another structure and make the callback set a flag there, then check the flag. Either way, it would need a retry loop that computes the remaining time.

For a service notification APC queued before CloseServiceHandle (e.g. if the service status changes immediately after the sleep times out), maybe it could be flushed by an alertable zero-duration sleep after CloseServiceHandle but before freeing the structure. But first it would be nice to know whether the risk actually exists; maybe CloseServiceHandle instead cancels the APC after all.

Reply:
From [SleepEx docs](https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-sleepex): 
> For the completion routine to be executed, the thread that called the I/O function must be in an alertable wait state when the completion callback function occurs. A thread goes into an alertable wait state by calling either SleepEx, [...], with the function's bAlertable parameter set to TRUE.

From [QueueUserAPC docs](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-queueuserapc):
> When a user-mode APC is queued, **the thread is not directed to call the APC function unless it is in an alertable state**. After the thread is in an alertable state, the thread handles all pending APCs [...]

My understanding is that APCs are not executed if the thread is not in an alertable state, but it is not clear if the alertable state persists after SleepEx returns. If that's not the case, I don't think that a stack corruption is possible, right?

Reply:
The alertable state won't persist after SleepEx returns, but I don't know whether Windows updates the SERVICE_NOTIFY structure even before it queues the APC to the thread.

Reply:
Ok, this is clear now. It would be very bad if they did that, because that couldn't be prevented even if the APC was canceled. Since the docs don't seem to help, how can we find that out?

Reply:
Perhaps easiest to use a user-mode debugger on a C++ test program. Set a data write breakpoint at _SERVICE_NOTIFY_2W::dwNotificationTriggered and a code execution breakpoint at the callback. Get stack traces of both, with correct Windows symbols. That should show which of these is the case:

* Windows has a user-mode APC routine that first updates _SERVICE_NOTIFY_2W and then calls the application's callback. The name of that routine should appear in both stack traces.
* Windows has a user-mode APC routine that updates _SERVICE_NOTIFY_2W and then queues a separate APC for the application's callback. This seems unlikely because it would cause more context switches than necessary.
* Windows has a kernel-mode APC routine or I/O completion routine that updates _SERVICE_NOTIFY_2W and then queues a user-mode APC for the application's callback. I think the data breakpoint would not even trigger.